File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Prototype/Prototype.ts
interface Addresses {
  city: string;
  country: string;
}
class User {
  public name: string;
  public address: Addresses;
  constructor(name: string, address: Addresses) {
    this.name = name;
    this.address = address;
  }
  public clone(): User {
    return new User(this.name, { ...this.address });
  }
}
const user:User = new User("mahdi", { city: "Minudasht", country: "Iran" });
const cloneUser = user.clone();
console.log("user:", user);
console.log("cloneUser:", cloneUser);
cloneUser.name = "ali";
cloneUser.address = { city: "Teheran", country: "Iran" };
console.log("user:(after)", user);
console.log("cloneUser:(after)", cloneUser);

File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Prototype/Prototype.py
from typing import Dict

# User address dataclass (for type hinting and immutability)
class Address:
    def __init__(self, city: str, country: str) -> None:
        self.city = city
        self.country = country


# User class with dataclass for type hinting and immutability
class User:
    def __init__(self, name: str, address: Address) -> None:
        self.name = name
        self.address = address

    def clone(self) -> "User":
        """Creates a deep copy of the User object."""
        return User(self.name, Address(self.address.city, self.address.country))


def main() -> None:
    prototype_user = User("Prototype User", Address("Prototype City", "Prototype Country"))
    cloned_user = prototype_user.clone()

    print("Prototype User:", prototype_user)
    print("Cloned User:", cloned_user)

    # Modifying the cloned user
    cloned_user.name = "Cloned User"
    cloned_user.address.city = "Cloned City"

    print("Prototype User (after modification):", prototype_user)
    print("Cloned User (after modification):", cloned_user)


if __name__ == "__main__":
    main()
